### PR TITLE
Adding some user warnings to potentially large problems or long runtimes

### PR DESCRIPTION
**Pull request by adelebai**
Adding some user warnings to potentially large problems or long runtimes.

The "large problem" definition is arbitrary and based on historical usage. 
Timeout threshold is also debatable and based on some historical data. However we can change them easily in the future.

### CODE REVIEW

File: azure-quantum/azure/quantum/optimization/problem.py [hunk 1]
```diff
@@ -212,3 +212,16 @@ def evaluate(self, configuration: Union[Dict[int, int], Dict[str, int]]) -> floa
...
+        "large" is an arbitrary threshold and can be easily changed. Based on usage data, we have defined a 
+        large problem to be 2500+ variables AND 1mil+ terms. 
+        """
+
+        set_vars = set()
+        num_terms = len(self.terms)
+        for term in self.terms:
+            set_vars.update(term.ids)
+        
+        return (len(set_vars) >= 2500 and len(self.terms) >= 10e6)
```
Comment (sanjgupt) [inline azure-quantum/azure/quantum/optimization/problem.py:227]: Can we make 2500 and 10e6 static constants?

File: azure-quantum/azure/quantum/optimization/solvers.py [hunk 2]
```diff
@@ -107,6 +107,24 @@ def optimize(self, problem: Union[str, Problem]):
...
             of a Problem has been uploaded.
         """
+        # print a warning if we suspect the job may take long based on its configured properties.
+        if not isinstance(problem, str):
+            is_large_problem = problem.is_large()
+            if is_large_problem:
+                if nested_params and "sweeps" in self.params["params"]:
+                    sweeps = self.params["params"]["sweeps"]
+                    # if problem is large and sweeps is large, warn. 
+                    if sweeps >= 10000:
```
Comment (sanjgupt) [inline azure-quantum/azure/quantum/optimization/solvers.py:117]: Same as above for hardcoded values

**Review by sanjgupt — APPROVED** [review]

File: azure-quantum/azure/quantum/optimization/problem.py [hunk 3]
```diff
@@ -212,3 +218,16 @@ def evaluate(self, configuration: Union[Dict[int, int], Dict[str, int]]) -> floa
                 total_cost += term.evaluate(configuration_transformed)
 
         return total_cost
+
+    def is_large(self) -> bool:
+        """Determines if the current problem is large. 
+        "large" is an arbitrary threshold and can be easily changed. Based on usage data, we have defined a 
+        large problem to be 2500+ variables AND 1mil+ terms.
```
Comment (vxfield) [inline azure-quantum/azure/quantum/optimization/problem.py:225]: can we refer the constants (`NUM_VARIABLES_LARGE ` and `NUM_TERMS_LARGE `) here ?

also, can we spell out `1mil+` to `1million+`, as in some contexts `mil` can mean `thousand`

**Review by vxfield — APPROVED** [review]
Approved with a couple comments

File: azure-quantum/azure/quantum/optimization/problem.py (hunk 1 above)
Comment (adelebai) [reply azure-quantum/azure/quantum/optimization/problem.py]: done

File: azure-quantum/azure/quantum/optimization/solvers.py (hunk 2 above)
Comment (adelebai) [reply azure-quantum/azure/quantum/optimization/solvers.py]: done

File: azure-quantum/azure/quantum/optimization/problem.py (hunk 3 above)
Comment (adelebai) [reply azure-quantum/azure/quantum/optimization/problem.py]: done

**Review by sanjgupt — CHANGES_REQUESTED** [review]
Actually I just realized logger.warn is deprecated and should be ideally replaces by logger.warning

**Comment by adelebai** [comment]
> Actually I just realized logger.warn is deprecated and should be ideally replaces by logger.warning

thanks for letting me know! I was in the process of testing this just now.

**Review by sanjgupt — APPROVED** [review]

**Pull request merged**